### DISCUSSION TITLE
Using FIPS provider without OpenSSL "enable-fips" config option

### DISCUSSION THREAD

**Discussion by jsrikanth1608** [openssl/openssl#24907] [Q&A] [answered]
Hello,

My understanding on usage of FIPS provider in the application needs to configure the OpenSSL with "enable-fips".  But in one of the use case the FIPS provider usage is determined at run time based on the provider module (fips.so) exist or not at specific location.
Please let me know, 
1. what is the behavior of the application using OpenSSL library which is not configured with "enable-fips" and using the FIPS provider ?
2. There are lot of references to FIPS_MODULE defines in the library, which are not enabled in this case, causes any "ABI or structure compatibility issues b/w OpenSSL and FIPS provider.?"

**Comment by t8m** [comment, marked as answer]
The `enable-fips` build time option just enables build of the fips.so provider module. The FIPS_MODULE defines are defined only when the fips.so is being built. It is undefined when libcrypto.so is being built. Thus the libcrypto will be identical regardless on whether `enable-fips` was specified or not.

You can even build the fips provider from a different version of the OpenSSL source release than the libcrypto and if you place the fips.so provider into the correct installation directory and properly configure it, it will work with the libcrypto from the original release. I.e., usually you want to use the latest release of openssl for the libcrypto/libssl build and the latest validated fips provider version build.

**Reply by jsrikanth1608** [reply]
Thank you @t8m for the details.

**Reply by p-rubrik** [reply]
But as part of documentation, we suggest to build latest openssl (used for libssl/libcrypto) with flag `-enable-fips` at https://github.com/openssl/openssl/blob/master/README-FIPS.md#download-and-build-the-latest-release-of-openssl

Does it have any significance for FIPS validation whether `-enable-fips` is provided or not in this step?

**Reply by paulidale** [reply]
No, you can build the latest version without the `enable-fips` option and use a FIPS provider from an earlier version.  This is an oversight in the documentation.

**Reply by slontis** [reply]
I assume this would also disable all the fips tests without the enable-fips option, so if you want to replace the fips provider and test then you would probably still want the enable-fips option.

**Comment by p-rubrik** [comment]
Hi Shane,
   This is about building the FIPS provider only. So there is no 'without
enable-fips' build configuration.

   We build OpenSSL/libssl/libcrypto without enable-fips options in a
different debian package.
   For this we use the debian build script directly without modification.

Thanks,
praveen
